src/Raven.Server/Rachis/ClusterMaintenanceMaster.cs

```diff
@@ -0,0 +1,317 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net.Sockets;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Raven.Client.Documents;
+using Raven.Client.Server;
+using Raven.Client.Server.Commands;
+using Raven.Client.Server.Tcp;
+using Raven.Server.Json;
+using Raven.Server.Utils;
+using Sparrow.Collections;
+using Sparrow.Collections.LockFree;
+using Sparrow.Json;
+using Sparrow.Logging;
+
+namespace Raven.Server.Rachis
+{
+    public class ClusterMaintenanceMaster : IDisposable
+    {
+        private readonly string _leaderClusterTag;
+
+        //maintenance handler is valid for specific term, otherwise it's requests will be rejected by nodes
+        private readonly long _term;
+        private bool _isDisposed;
+
+        private readonly ConcurrentDictionary<string, ClusterNode> _clusterNodes = new ConcurrentDictionary<string, ClusterNode>();
+        private readonly ConcurrentDictionary<ClusterNode, long> _failedNodes = new ConcurrentDictionary<ClusterNode, long>();
+        private readonly ConcurrentDictionary<ClusterNode, long> _timedOutNodes = new ConcurrentDictionary<ClusterNode, long>();
+
+        private readonly CancellationTokenSource _cts = new CancellationTokenSource();
+
+        private readonly JsonContextPool _contextPool = new JsonContextPool();
+
+        public ClusterMaintenanceMaster(string leaderClusterTag, long term)
+        {
+            _leaderClusterTag = leaderClusterTag;
+            _term = term;
+        }
+
+        public async Task AddToCluster(string clusterTag, string url)
+        {
+            var connectionInfo = await ReplicationUtils.GetTcpInfoAsync(MultiDatabase.GetRootDatabaseUrl(url), null, null);
+            var clusterNode = new ClusterNode(clusterTag, connectionInfo, _contextPool, this, _cts.Token);
+            _clusterNodes.Add(clusterTag, clusterNode);
+            var task = clusterNode.StartListening();
+            GC.KeepAlive(task); // we are explicitly not waiting on this task
+        }
+
+        public Dictionary<string, ClusterNodeStatusReport> GetStats()
+        {
+            var clusterStats = new Dictionary<string, ClusterNodeStatusReport>();
+            foreach (var node in _clusterNodes)
+            {
+                clusterStats[node.Key] = node.Value.ReceivedReport;
+            }
+            return clusterStats;
+        }
+
+        public void RemoveFromCluster(string clusterTag)
+        {
+            _clusterNodes[clusterTag]?.Dispose();
```

Comment:
Bad idea. If the value doesn't exists, you'll throw.
You are keeping a disposed instance in the dictionary, etc.

Better to use `TryRemove` and then dispose it.